index.js

```diff
@@ -113,66 +113,103 @@ class CarmiObserver extends React.Component {
 CarmiObserver.contextType = CarmiContext;
 
 function Provider({ children, value, compsLib = {} }) {
-  const privates = getPrivates(value);
+  const privates = getPrivatesByInstance(value);
   privates.compsLib = compsLib;
   return React.createElement(CarmiRoot, { children, value });
 }
 
-
 function getMaybeKey(props, name) {
   return props && props.hasOwnProperty(name) ? props[name] : null;
 }
 
-function createElement(descriptor) {
+function replaceElementProps(element, newProps) {
+  Object.assign(element.props, newProps);
+  Object.keys(element.props).forEach(prop => {
+    if (!newProps.hasOwnProperty(prop) && !BUILT_IN_PROPS.hasOwnProperty(prop)) {
+      delete element.props[prop];
+    }
+  });
+}
+
+function parseDescriptor(descriptor) {
   const type = descriptor[0];
-  const childProps = descriptor[1] || {};
-  const { key: rawKey, ...childExtraProps } = childProps;
-  const key = getMaybeKey(childProps, 'key');
-  const privates = getPrivates(this);
-  const prevElement = privates.descriptorToElementsMap.get(descriptor);
-  if (prevElement && prevElement.props.type === type && getMaybeKey(prevElement.props, 'origKey') === key) {
+  const props = descriptor[1] || {};
+  const children = descriptor.slice(2);
+  const childrenList = children && Array.isArray(children[0]) && children.length === 1 ? children[0] : children;
+  const { key: rawKey, ...extraProps } = props;
+  const key = getMaybeKey(props, 'key');
+  return {type, props, children, childrenList, rawKey, extraProps, key};
+}
+
+const wrapElement = (wrappers, element, children) => wrappers.reduce((wrappedElement, wrapper) => wrapper(wrappedElement, children), element)
+
+function createElement(wrappers, descriptor) {
+  const {type, props, childrenList, extraProps, key} = parseDescriptor(descriptor);
+  const privates = getPrivatesByInstance(this);
+  const currentElement = privates.descriptorToElementsMap.get(descriptor);
+  if (currentElement && currentElement.props.type === type && getMaybeKey(currentElement.props, 'origKey') === key) {
+    // Element is mounted
     if (privates.root && privates.descriptorToCompsMap.has(descriptor)) {
       privates.descriptorToCompsMap.get(descriptor).forEach(comp => privates.pendingFlush.add(comp));
     } else {
-      prevElement.props.dirtyFlag[0] = true;
+      currentElement.props.dirtyFlag[0] = true;
     }
-    Object.assign(prevElement.props, childExtraProps);
-    Object.keys(prevElement.props).forEach(prop => {
-      if (!childExtraProps.hasOwnProperty(prop) && !BUILT_IN_PROPS.hasOwnProperty(prop)) {
-        delete prevElement.props[prop];
-      }
-    });
+    replaceElementProps(currentElement, extraProps);
   } else {
     const dirtyFlag = [true];
-    const props = { descriptor, type, dirtyFlag };
+    const observerProps = {descriptor, type, dirtyFlag, token: pointerByInstance.get(this)};
```

Comment:
getting the token directly from the pointerByInstance makes an assumption that getPrivatesByInstance was called before. Probably better to create a dedicated getter that does not make this assumption.